### PR TITLE
Add http3 support to newest nginx

### PR DESCRIPTION
**Pull request by bastakka**
I found this patch online which makes the build possible even on the newest nginx version.

### CODE REVIEW

**Comment by GwynethLlewelyn** [comment]
Awesome! This fixes #195 for me! Thank you so much!

**Comment by MaximeMichaud** [comment]
I didn't test everything, but yes compilation seem to be fixed. I still need to test if HTTP3 work correctly or not, but I don't see why it would not.

**Comment by GwynethLlewelyn** [comment]
I'd say it sort of works... but, then again, that has been my overall experience with HTTP/3. It's very picky; sometimes it works, sometimes it refuses to work, sometimes a refresh helps, sometimes _nothing_ helps, sometimes switching browsers do the trick, and sometimes not even that. But that's what I always get from my setup, so I guess it's the way it is.

I keep wondering how Cloudflare, Google and others are able to keep their HTTP/3 connections rock-solid and very predictable...

**Comment by MaximeMichaud** [comment]
> I'd say it sort of works... but, then again, that has been my overall experience with HTTP/3. It's very picky; sometimes it works, sometimes it refuses to work, sometimes a refresh helps, sometimes _nothing_ helps, sometimes switching browsers do the trick, and sometimes not even that. But that's what I always get from my setup, so I guess it's the way it is.
> 
> I keep wondering how Cloudflare, Google and others are able to keep their HTTP/3 connections rock-solid and very predictable...

They have been working on it since a long time and the support is well done I guess. Right there, we are only using a stupid patch. I'm still optimistic for OPENSSL & Official NGINX Support. I do hope this patch work correctly.

**Comment by bastakka** [comment]
I have a similar experience with http3 as Gwyneth, but with this, I was able to at least run the latest Nginx with the http3 module installed and with successful quic connections. Nginx-quic repository would probably be better for that, but I did not get much better results with it anyway. This way, you can install it with this script and have no worries about manually compiling boringssl and building Nginx-quic.

**Comment by bastakka** [comment]
It would maybe be much better if the patch itself would be added to this repo instead of being downloaded from another repo that has more files than needed. I just wanted to point out the existence of this patch. This pull would be reckless. I will try to make a better one

**Comment by bastakka** [comment]
This should do the job.

File: nginx-autoinstall.sh [hunk 1]
```diff
@@ -99,7 +99,7 @@ case $OPTION in
 		echo ""
 		echo "Modules to install :"
 		while [[ $HTTP3 != "y" && $HTTP3 != "n" ]]; do
-			read -rp "       HTTP/3 (⚠️ Patch by Cloudflare for versions <= 1.19.7, will install BoringSSL, Quiche, Rust and Go) [y/n]: " -e -i n HTTP3
+			read -rp "       HTTP/3 (⚠️ Patch by Cloudflare , will install BoringSSL, Quiche, Rust and Go) [y/n]: " -e -i n HTTP3
```
Comment (bastakka) [inline nginx-autoinstall.sh:102]: Remove version warning from http3 install prompt

File: nginx-autoinstall.sh [hunk 2]
```diff
@@ -510,7 +510,6 @@ case $OPTION in
 	if [[ $HTTP3 == 'y' ]]; then
 		cd /usr/local/src/nginx/modules || exit 1
 		git clone --depth 1 --recursive https://github.com/cloudflare/quiche
-		git clone https://github.com/patrikjuvonen/docker-nginx-http3.git
```
Comment (bastakka) [inline nginx-autoinstall.sh]: Remove cloning of not needed repo

File: nginx-autoinstall.sh [hunk 3]
```diff
@@ -520,7 +519,10 @@ case $OPTION in
 		cd /usr/local/src/nginx/nginx-${NGINX_VER} || exit 1
 		# Apply actual patch
 		patch -p01 </usr/local/src/nginx/modules/quiche/extras/nginx/nginx-1.16.patch
-		patch -p01 </usr/local/src/nginx/modules/docker-nginx-http3/nginx-1.19.7.patch
```
Comment (bastakka) [inline nginx-autoinstall.sh]: Move patch to right place

File: nginx-autoinstall.sh [hunk 4]
```diff
@@ -520,7 +519,10 @@ case $OPTION in
...
+
+		# Apply patch for nginx > 1.19.7
+		wget https://raw.githubusercontent.com/angristan/nginx-autoinstall/master/patches/nginx-http3-1.19.7.patch -O nginx-http3.patch
+		patch -p01 <nginx-http3.patch
```
Comment (bastakka) [inline nginx-autoinstall.sh:525]: Download and install patch

**Review by bastakka — COMMENTED** [review]
Explanation of what I changed from my first PR.

**Review by bastakka — COMMENTED** [review]
Add the needed patch to this repo

**Comment by MaximeMichaud** [comment]
> I have a similar experience with http3 as Gwyneth, but with this, I was able to at least run the latest Nginx with the http3 module installed and with successful quic connections. Nginx-quic repository would probably be better for that, but I did not get much better results with it anyway. This way, you can install it with this script and have no worries about manually compiling boringssl and building Nginx-quic.

So, do you recommend using it in production ? In my opinion, I will probably wait for a better support if these something that may not work properly.

**Comment by bastakka** [comment]
> So, do you recommend using it in production ? In my opinion, I will probably wait for a better support if these something that may not work properly.

I cannot say I would not recommend it. This way you add http3 support to your server, which can work, and in times where it does not, your browser will always use http2 which will work every time. I have never experienced my page not loading, just not with http3 every time. To be honest I do not even think http3 is needed in production yet. Http2 does work these days well enough.

**Comment by MaximeMichaud** [comment]
> > So, do you recommend using it in production ? In my opinion, I will probably wait for a better support if these something that may not work properly.
> 
> I cannot say I would not recommend it.
> 
> This way you add http3 support to your server, which can work, and in times where it does not, your browser will always use http2 which will work every time. I have never experienced my page not loading, just not with http3 every time.
> 
> To be honest I do not even think http3 is needed in production yet. Http2 does work these days well enough.

OK perfect, if the browser still use HTTP2 in fallback and the page loading, that not so bad.
Thank you,

**Comment by GwynethLlewelyn** [comment]
I've been using it in production for quite a while now. Even though I'm not really running mission-critical applications (say, controlling NASA satellites or the New York Stock Exchange...), I'm confident enough that this patch is working _very_ well with all my currently deployed websites.

Aye, it's true that sometimes browsers are unable to make an HTTP/3 connection (and fall back to HTTP/2), and, as of today, I cannot understand why not. But this is the behaviour I have always had with whatever HTTP/3 module was compiled into `nginx`. I'm always surprised that Cloudflare and Facebook which use `nginx` internally can get HTTP/3 working _all the time_. I guess it has to do with something at the networking layer.

On the other hand, when testing with `cURL` as opposed to using one of the many browsers I use for testing (Brave, Firefox, Safari, Edge...), it always gets a correct HTTP/3 connection. This _might_ be because my home-compiled `cURL` _also_ uses Quiche. There _might_ be an issue connecting Quiche and non-Quiche endpoints (but, as said, anything behind Cloudflare's Quiche-enabled `nginx` has _no_ issues with _either_ browser). Note that the same reasoning applies to using BoringSSL instead of OpenSSL — if it works for Cloudflare, it ought to work for us as well.

**Comment by GwynethLlewelyn** [comment]
Perhaps even more important than fixing HTTP/3, this patch _also_ fixes issue #195, which is _still_ present in the main branch...

File: nginx-autoinstall.sh (hunk 1 above)
Comment (angristan) [inline nginx-autoinstall.sh:102]:
```suggestion
			read -rp "       HTTP/3 (⚠️ Patch by Cloudflare, will install BoringSSL, Quiche, Rust and Go) [y/n]: " -e -i n HTTP3
```

File: README.md [hunk 5]
```diff
@@ -22,7 +22,7 @@ The script might work on ARM-based architectures, but it's only being regularly
 
 - [LibreSSL from source](http://www.libressl.org/) (CHACHA20, ALPN for HTTP/2, X25519, P-521)
 - [OpenSSL from source](https://www.openssl.org/) (TLS 1.3, CHACHA20, ALPN for HTTP/2, X25519, P-521)
-- [Cloudflare's patch for HTTP/3](https://blog.cloudflare.com/experiment-with-http-3-using-nginx-and-quiche/) with [Quiche](https://github.com/cloudflare/quiche) and [BoringSSL](https://github.com/google/boringssl). (⚠️ the patch [doesn't work for versions > 1.19.6](https://github.com/cloudflare/quiche/issues/859).)
+- [Cloudflare's patch for HTTP/3](https://blog.cloudflare.com/experiment-with-http-3-using-nginx-and-quiche/) with [Quiche](https://github.com/cloudflare/quiche) and [BoringSSL](https://github.com/google/boringssl).
```
Comment (angristan) [inline README.md:25]:
```suggestion
- [Cloudflare's patch for HTTP/3](https://blog.cloudflare.com/experiment-with-http-3-using-nginx-and-quiche/) with [Quiche](https://github.com/cloudflare/quiche) and [BoringSSL](https://github.com/google/boringssl).
```

**Comment by angristan** [comment]
Thanks everyone.

@bastakka In the original post you said:
> I found this patch online

Could we have the source please?

File: nginx-autoinstall.sh [hunk 6]
```diff
@@ -520,6 +520,10 @@ case $OPTION in
 		# Apply actual patch
 		patch -p01 </usr/local/src/nginx/modules/quiche/extras/nginx/nginx-1.16.patch
 
+		# Apply patch for nginx > 1.19.7
```
Comment (angristan) [inline nginx-autoinstall.sh:523]:
```suggestion
		# Apply patch for nginx > 1.19.7 (source: https://github.com/cloudflare/quiche/issues/936#issuecomment-857618081)
```

**Review by angristan — APPROVED** [review]

**Comment by MaximeMichaud** [comment]
I did have some issues recently, like bastakka , I will not use it in production.
2021/09/17 06:56:39 [alert] 2121#2121: *24 epoll_ctl(1, 12) failed (17: File exists), client
2021/09/17 06:57:24 [alert] 2121#2121: *21 sendmsg() failed (9: Bad file descriptor) while writing QUIC packets, client: 93.10.4.226, server: 0.0.0.0:443

**Comment by bastakka** [comment]
I do not have a problem like this. I mentioned that I do not load pages via h3 every time. I would suggest trying to change events from epoll to poll if that helps, but I do not have any troubleshooting to your problem.

**Comment by GwynethLlewelyn** [comment]
Let me echo @MaximeMichaud, I also have precisely the same issue. What seems to be happening is that the browser attempts to do an HTTP/3 connection, gets a `500` error, tries a bit more, and gives up, using HTTP/2 from then on (obviously, you can clean browser caches and so forth in order to try again). The errors are as @MaximeMichaud describe.

However, I _might_ have a very slight clue about what causes this: the Google PageSpeed module (which I have installed and activated practically everywhere). Here is why I think that: _usually_, on most HTTP connections, what I see in the logs is that the request goes to nginx first, which passes it along to the PageSpeed module; if it's cached there, it's served directly (no logs); but if it isn't, PageSpeed _seems_ to make a _new_ call to nginx (meaning: the _origin_ of the call is the same host where the nginx server is running) to retrieve the resource first and you get an _extra_ entry on the logs. I don't know if this is the intended behaviour or just some weird configuration setting that I have placed somewhere, not really knowing why I did that...

My wild guess is that this interferes somehow with HTTP/3, but I cannot really say _why_ it should — after all, it's almost certain that Google uses PageSpeed with their own internal web server solution, which supports HTTP/3. So it _could_ be just a configuration error at some point; PageSpeed has a gazillion options, and it's not impossible that hidden deep in those layers and layers of options is a special setting to get it to work flawlessly with HTTP/3. The truth is that I haven't tried out my assumption...

Anyway, a few more logs, just to give you some context:

`my.server.tld` is the host running nginx (let's assume its IPv4 address is `10.2.3.4`).
`10.20.30.40` is the IPv4 address dynamically assigned to my router at home.

Here's the full nginx configuration:

```sh
# nginx -V
nginx version: nginx/1.21.2
built by gcc 9.3.0 (Ubuntu 9.3.0-17ubuntu1~20.04)
built with OpenSSL 1.1.1 (compatible; BoringSSL) (running with BoringSSL)
TLS SNI support enabled
configure arguments: --prefix=/etc/nginx --sbin-path=/usr/sbin/nginx --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --user=nginx --group=nginx --with-cc-opt=-Wno-deprecated-declarations --with-cc-opt=-Wno-ignored-qualifiers --with-openssl=/usr/local/src/nginx/modules/quiche/deps/boringssl --with-quiche=/usr/local/src/nginx/modules/quiche --with-http_v2_hpack_enc --with-threads --with-file-aio --with-http_ssl_module --with-http_v2_module --with-http_mp4_module --with-http_auth_request_module --with-http_slice_module --with-http_stub_status_module --with-http_realip_module --with-http_sub_module --add-module=/usr/local/src/nginx/modules/incubator-pagespeed-ngx-1.13.35.2-stable --add-module=/usr/local/src/nginx/modules/ngx_brotli --add-module=/usr/local/src/nginx/modules/ngx_cache_purge --with-http_v3_module
```
My test apps are [Grav](https://getgrav.org/) (a flat-file CMS) v1.7.20, and WordPress; I'm showing only the logs for Grav below.

Here are the contents of `error.log`:
```
2021/09/13 15:55:56 [alert] 1149317#1149317: *19081 epoll_ctl(1, 118) failed (17: File exists), client: 10.20.30.40 server: my.server.tld, request: "POST /admin HTTP/3", host: "my.server.tld", referrer: "https://my.server.tld/admin"
2021/09/13 15:56:27 [alert] 1149309#1149309: *19281 epoll_ctl(1, 102) failed (17: File exists), client: 10.20.30.40, server: my.server.tld, request: "POST /admin HTTP/3", host: "my.server.tld", referrer: "https://my.server.tld/admin"
```
... and the equivalent contents of `access.log`:

```
10.20.30.40 - - [13/Sep/2021:15:55:56 +0100] "POST /admin HTTP/3" 500 1893 "https://my.server.tld/admin" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.63 Safari/537.36"
10.20.30.40 - - [13/Sep/2021:15:56:27 +0100] "POST /admin HTTP/3" 500 1893 "https://my.server.tld/admin" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.63 Safari/537.36"
```

I would guess that @MaximeMichaud has similar logs to those above.

Unfortunately, there is nothing more on the logs...

**Comment by MaximeMichaud** [comment]
No PageSpeed on my end.
configure arguments: --prefix=/etc/nginx --sbin-path=/usr/sbin/nginx --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --user=nginx --group=nginx --with-cc-opt=-Wno-deprecated-declarations --with-cc-opt=-Wno-ignored-qualifiers --with-openssl=/usr/local/src/nginx/modules/quiche/deps/boringssl --with-quiche=/usr/local/src/nginx/modules/quiche --with-threads --with-file-aio --with-http_ssl_module --with-http_v2_module --with-http_mp4_module --with-http_auth_request_module --with-http_slice_module --with-http_stub_status_module --with-http_realip_module --with-http_sub_module --add-module=/usr/local/src/nginx/modules/ngx_brotli --add-module=/usr/local/src/nginx/modules/headers-more-nginx-module-0.33 --add-module=/usr/local/src/nginx/modules/ModSecurity-nginx --with-http_v3_module

I cannot describe why this is happening.
Just login on WordPress (example) seem to trigger the problem, but this is not happening for everybody.

I cannot wait for the official patch.
https://hg.nginx.org/nginx-quic/

**Comment by mtx-z** [comment]
Same issue here: all POST requests on HTTP/3 are failing. This is a [known issue](https://giters.com/cloudflare/quiche/issues/640).

However, the [README](https://hg.nginx.org/nginx-quic/file/tip/README) [now indicates](https://hg.nginx.org/nginx-quic/rev/98e94553ae51) (since 12/09/22, yesterday as of today) that there are some production usages:
image


Logs:
```
2022/09/13 00:18:13 [alert] 2392567#2392567: *70 epoll_ctl(1, 12) failed (17: File exists), client: 111.111.111.111, server: klesiatest.mtx.io, request: "POST /wp-admin/admin-ajax.php HTTP/3", host: "domain.com", referrer: "https://domain.com/wp-admin/admin.php?..."
2022/09/13 00:20:10 [alert] 2392570#2392570: *71 epoll_ctl(1, 17) failed (17: File exists), client: 111.111.111.111, server: klesiatest.mtx.io, request: "POST /wp-admin/admin-ajax.php HTTP/3", host: "domain.com", referrer: "https://domain.com/wp-admin/admin.php?..."
```


Versions:
```
root@VM-TEST:/# nginx -V
nginx: /usr/local/lib/libcurl.so.4: no version information available (required by /usr/local/modsecurity/lib/libmodsecurity.so.3)
nginx version: nginx/1.22.0
built by gcc 10.2.1 20210110 (Debian 10.2.1-6)
built with OpenSSL 1.1.1 (compatible; BoringSSL) (running with BoringSSL)
TLS SNI support enabled
configure arguments: --prefix=/etc/nginx --sbin-path=/usr/sbin/nginx --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --pid-path=/var/run/nginx.pid --lock-path=/var/run/nginx.lock --http-client-body-temp-path=/var/cache/nginx/client_temp --http-proxy-temp-path=/var/cache/nginx/proxy_temp --http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp --user=nginx --group=nginx --with-cc-opt=-Wno-deprecated-declarations --with-cc-opt=-Wno-ignored-qualifiers --with-openssl=/usr/local/src/nginx/modules/quiche/quiche/deps/boringssl --with-quiche=/usr/local/src/nginx/modules/quiche --with-http_v2_hpack_enc --with-threads --with-file-aio --with-http_ssl_module --with-http_v2_module --with-http_mp4_module --with-http_auth_request_module --with-http_slice_module --with-http_stub_status_module --with-http_realip_module --with-http_sub_module --with-http_flv_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_secure_link_module --with-stream --with-stream_ssl_module --with-ipv6 --with-http_perl_module --with-compat --with-http_addition_module --with-mail --with-mail_ssl_module --with-stream_realip_module --with-stream_ssl_preread_module --with-http_random_index_module --add-module=/tmp/ansible-vps-php-webapp-stack/nginx/modules/nginx-length-hiding-filter-module --add-module=/usr/local/src/nginx/modules/incubator-pagespeed-ngx-1.13.35.2-stable --add-module=/usr/local/src/nginx/modules/ngx_brotli --add-module=/usr/local/src/nginx/modules/headers-more-nginx-module-0.33 --add-module=/usr/local/src/nginx/modules/ngx_cache_purge --add-module=/usr/local/src/nginx/modules/ngx_http_substitutions_filter_module --add-module=/usr/local/src/nginx/modules/fancyindex --with-http_dav_module --add-module=/usr/local/src/nginx/modules/nginx-dav-ext-module --add-module=/usr/local/src/nginx/modules/nginx-module-vts --add-module=/usr/local/src/nginx/modules/nginx-rtmp-module --add-module=/usr/local/src/nginx/modules/testcookie-nginx-module --add-module=/usr/local/src/nginx/modules/ModSecurity-nginx --with-http_v3_module
root@VM-TEST:/# curl -V
curl 7.85.0-DEV (x86_64-pc-linux-gnu) libcurl/7.85.0-DEV OpenSSL/3.0.5 zlib/1.2.12 brotli/1.0.9 ngtcp2/0.9.0-DEV nghttp3/0.8.0-DEV
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli HSTS HTTP3 HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL threadsafe TLS-SRP UnixSockets
```

**Comment by GwynethLlewelyn** [comment]
@mtx-z I did, even though my "real case" websites are all very, very low traffic :-)

nginx version: nginx/1.23.1
built by gcc 11.2.0 (Ubuntu 11.2.0-19ubuntu1)
built with OpenSSL 3.0.5+quic 5 Jul 2022
TLS SNI support enabled
configure arguments: --with-ld-opt=-static-libstdc++ --add-module=/usr/local/src/nginx/modules/incubator-pagespeed-ngx [...]

It certainly works. **Much** better than BoringSSL + Cloudflare Quiche!

**Comment by MaximeMichaud** [comment]
Good news finally !

Changes with nginx 1.25.1                                        13 Jun 2023

    *) Feature: the "http2" directive, which enables HTTP/2 on a per-server
       basis; the "http2" parameter of the "listen" directive is now
       deprecated.

    *) Change: HTTP/2 server push support has been removed.

    *) Change: the deprecated "ssl" directive is not supported anymore.

    *) Bugfix: in HTTP/3 when using OpenSSL.

Changes with nginx 1.25.0                                        23 May 2023

    *) Feature: experimental HTTP/3 support.

**Pull request merged**